src/d3a/d3a_core/sim_results/kpi.py

```diff
@@ -0,0 +1,147 @@
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer_wh += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list:
+            # self_consumption_buffer needs to be exhausted to total_self_consumption
+            # if sold to internal consumer
+            if trade.buyer_origin in self.consumer_list and self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                    self.total_self_consumption_wh += trade.offer.energy * 1000
+                else:
+                    self.total_self_consumption_wh += self.self_consumption_buffer_wh
+                    self.self_consumption_buffer_wh = 0
+            # self_consumption_buffer needs to be exhausted if sold to any external agent
+            elif trade.buyer_origin not in [*self.ess_list, *self.consumer_list] and \
+                    self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                else:
+                    self.self_consumption_buffer_wh = 0
+
+    def _accumulate_infinite_consumption(self, trade):
+        if trade.seller_origin in self.buffer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _dissipate_infinite_consumption(self, trade):
+        if trade.buyer_origin in self.buffer_list and trade.seller_origin in self.producer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def accumulate_energy_trace(self):
+        for c_area in self.consumer_area_list:
+            for market in c_area.past_markets:
+                for trade in market.trades:
+                    self._accumulate_self_consumption(trade)
+                    self._accumulate_self_production(trade)
+                    self._accumulate_self_consumption_buffer(trade)
+                    self._dissipate_self_consumption_buffer(trade)
+                    self._accumulate_infinite_consumption(trade)
+                    self._dissipate_infinite_consumption(trade)
+
+
+class KPI:
+    def __init__(self):
+        self.performance_indices = dict()
+        self.state = None
+
+    def __repr__(self):
+        return f"KPI: {self.performance_indices}"
+
+    def area_performance_indices(self, area):
+        self.state = KPIState()
+
+        self.state.accumulate_devices(area)
```

Comment:
I would expose one single `update(area)` method on the KPIState class, that would be only responsible for calling the `accumulate_devices` and `accumulate_energy_trace` methods. That way there is only one call for updating the statistics in the KPIState class.